addons/website_sale/controllers/main.py

```diff
@@ -1786,6 +1787,18 @@ def shop_confirm_order(self, **post):
 
         order_sudo._recompute_taxes()
         order_sudo._recompute_prices()
+
+        not_for_sale_products = []
+        for product in order_sudo.order_line.product_id:
+            if product == order_sudo.carrier_id.product_id:
+                continue
+            if not product._is_add_to_cart_allowed():
+                not_for_sale_products.append(product.name)
+        if len(not_for_sale_products) > 0:
+            error_product_not_for_sale_message = ''
+            for product_name_error in not_for_sale_products:
+                error_product_not_for_sale_message += (", " + product_name_error)
+            raise UserError("The following product(s) are no longer available" + error_product_not_for_sale_message)
```

Comment:
This can be expressed in a much more pythonic way relying on the fact that `order_sudo.order_line.product_id` is an iterable on which you can perform both `filtered` and `map`.

```suggestion
unavailable_products = order_sudo.order_line.product_id.filtered(lambda p: p != order_sudo.carrier_id.product_id and not p._is_add_to_cart_allowed())
        if unavailable_products:
            raise UserError("The following product(s) are no longer available" + ", ".join(unavailable_products.mapped('name')))
```